mgr: redirect python stdout,stderr to ceph log

tchaikov opened the pull request:
Signed-off-by: Kefu Chai <kchai@redhat.com>

tserong commented:
Good idea (this might also help with some CherryPy logging annoyance I've been looking at). I'll do some testing...

tserong commented:
OK, this works for me (the errors that weren't being logged that I mentioned in #14126 do get logged with this patch applied), and actually means we could drop the `handle_pyerror()` function and use  `PyErr_Print()` in `PyModules::init()` instead. Also, the CherryPy logging now gets captured without any extra effort.

tserong commented on src/mgr/PyModules.cc:
```diff
@@ -29,6 +29,34 @@
 #define dout_subsys ceph_subsys_mgr
+
+#undef dout_prefix
+#define dout_prefix *_dout << "mgr [py] "
```
I'd suggest this be "mgr[$module_name]" if possible, rather than "mgr [py]", so you know which module the log/error came from.

tchaikov replied:
@tserong i guess it's possible by parsing the backtrace but that'd terrify me.

tchaikov commented:
@tserong i removed the `handle_pyerror()` related changes from your PR and incorporated it into mine. i think that would form a more complete solution for debugging the ceph-mgr plugins.

tchaikov commented:
@tserong i removed the `handle_pyerror()` related changes from your PR and incorporated it into mine. i think that would form a more complete solution for debugging the ceph-mgr plugins.

tchaikov commented:
retest this please.

tserong commented:
Thanks!

tchaikov commented:
@tserong i updated the my change so that only the customized "stderr" is used only when the ceph-mgr is daemonized. the format of the original error message looks better without the interleaved new lines.

tchaikov commented:
now the output with "-f" looks like 

```
2017-03-29 15:32:09.473607 7f29a9180700 -1 mgr init Traceback (most recent call last):

  File "/var/ceph/ceph/src/pybind/mgr/rest/__init__.py", line 2, in <module>
    from module import *  # NOQA

  File "/var/ceph/ceph/src/pybind/mgr/rest/module.py", line 23, in <module>
    import cherrypy

ImportError: No module named cherrypy

2017-03-29 15:32:13.913007 7f29ac186700 -1 mgr ms_dispatch mgrdigest v1
2017-03-29 15:32:18.913380 7f29ac186700 -1 mgr ms_dispatch mgrdigest v1
2017-03-29 15:32:23.913688 7f29ac186700 -1 mgr ms_dispatch mgrdigest v1
Traceback (most recent call last):
  File "/var/ceph/ceph/src/pybind/mgr/fsstatus/module.py", line 267, in handle_command
2017-03-29 15:32:24.380101 7f29a897f700 -1 mgr handle_command Failed to invoke handle_command() on fsstatus
    return self.handle_osd_perf(cmd)
  File "/var/ceph/ceph/src/pybind/mgr/fsstatus/module.py", line 258, in handle_osd_perf
    raise Exception("omg")
Exception: omg
```

// i added a `raise Exception("omg")` in `fsstatus/module.py`

tchaikov commented:
retest this please.

jcsp commented:
(only just saw this one)

For the exception handling, I think it's cleaner to keep the explicit code that generates a string and sends it to the log, rather than relying on plumbing python's stderr into our log at level 0.

The problem with sending stderr to level 0 is that if you instantiate a python StreamHandler log handler, that goes to stderr by default.  So any module that happens to spin one of those up by default (like a web server that thinks its logging to the console) will end up generating level 0 messages in the ceph log.

So I guess I would suggest keeping the explicit exception handler (sending its strings to dout(0)), and maybe have the python stderr handler send its strings to dout(4) or something like that, so that if someone wants to see their python module stderr they can, but we don't send it to our log level 0 by default.

dmick commented:
wish I'd realized this was going on before I modified everything to use handle_pyerror() :)  But given the above comments, if the extra on top of my dmick/wip-guilolz branch is useful, I'm happy to contribute it, particularly https://github.com/dmick/ceph/commit/23cb9a7178965553fbcbe6be3ca5cf4a20de7a2f, https://github.com/dmick/ceph/commit/217f3f2b8b66296f5a3ce2391fd3be03653cb5e4, and https://github.com/dmick/ceph/commit/a3c3b4e53bf513e72a94257a83783f61bef771a2

yuriw commented:
Ran tests holding merge per @liewegas 
http://pulpito.front.sepia.ceph.com:80/yuriw-2017-04-12_22:08:16-rados-wip-yuri-testing_2017_4_13-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-04-13_04:53:51-rados-wip-yuri-testing_2017_4_13---basic-smithi/

tchaikov commented:
@jcsp

- keep `handle_pyerror()` for printing the exception when loading a plugin.
- drop the 3rd commit which replaced the `handle_pyerror()` call with `PyErr_Print()`.

could you take a look again?

jcsp commented on src/mgr/PyModules.h:
```diff
@@ -27,7 +27,9 @@ class ServeThread;
 
 class PyModules
 {
-  protected:
```
Why keep a reference to py_logger if we're not doing any teardown on it?

tchaikov replied:
right. will remove it.

liewegas commented:
rebase?

tchaikov commented:
@liewegas rebased

@jcsp updated as you suggested.

tchaikov commented:
http://pulpito.ceph.com/kchai-2017-04-21_03:04:04-rados-wip-kefu-testing---basic-mira
